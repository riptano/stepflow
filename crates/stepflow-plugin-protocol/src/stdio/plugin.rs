use error_stack::ResultExt;
use stepflow_plugin::{Plugin, PluginError, Result};
use stepflow_protocol::component_info::ComponentInfo;
use stepflow_workflow::{Component, Value};

use super::ClientHandle;

pub struct StdioPlugin {
    client: ClientHandle,
}

impl StdioPlugin {
    pub fn new(client: ClientHandle) -> Self {
        Self { client }
    }
}

impl Plugin for StdioPlugin {
    async fn init(&self) -> Result<()> {
        self.client
            .request(&stepflow_protocol::initialization::Request {
                runtime_protocol_version: 1,
            })
            .await
            .change_context(PluginError::Initializing)?;

        self.client
            .notify(&stepflow_protocol::initialization::Complete {})
            .await
            .change_context(PluginError::Initializing)?;

        Ok(())
    }

    async fn component_info(&self, component: &Component) -> Result<ComponentInfo> {
        // TODO: Enrich this? Component not found, etc. based on the protocol error code?
        let response = self
            .client
            .request(&stepflow_protocol::component_info::Request {
                component: component.clone(),
            })
            .await
            .change_context(PluginError::ComponentInfo)?;

        Ok(response.info)
    }

    async fn execute(&self, component: &Component, input: Value) -> Result<Value> {
        let response = self
            .client
            .request(&stepflow_protocol::component_execute::Request {
                component: component.clone(),
                input,
            })
            .await
            .change_context(PluginError::Execution)?;

        Ok(response.output)
    }
}
