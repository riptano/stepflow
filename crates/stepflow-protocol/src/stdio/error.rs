use std::path::PathBuf;

use thiserror::Error;

#[derive(Error, Debug)]
pub enum StdioError {
    #[error("error spawning stepflow component process")]
    Spawn,
    #[error("error sending message")]
    Send,
    #[error("error receiving message")]
    Recv,
    #[error("received invalid message")]
    InvalidMessage,
    #[error("received invalid response")]
    InvalidResponse,
    #[error("components server error({code}): {message}")]
    ServerError {
        code: i64,
        message: String,
        data: Option<serde_json::Value>,
    },
    #[error("components server failed with exit code {exit_code:?}")]
    ServerFailure { exit_code: Option<i32> },
    #[error("error closing stepflow component process")]
    Close,
    #[error("invalid command: {}", .0.display())]
    InvalidCommand(PathBuf),
    #[error("error in receive loop")]
    RecvLoop,
    #[error("command not found: {0}")]
    MissingCommand(String),
    #[error("unknown method: {method}")]
    UnknownMethod { method: String },
}

pub type Result<T, E = error_stack::Report<StdioError>> = std::result::Result<T, E>;
