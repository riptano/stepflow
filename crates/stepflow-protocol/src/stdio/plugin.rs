use std::sync::Arc;

use error_stack::ResultExt as _;
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use stepflow_core::{
    FlowResult,
    component::ComponentInfo,
    workflow::{Component, ValueRef},
};
use stepflow_plugin::{
    Context, DynPlugin, ExecutionContext, Plugin, PluginConfig, PluginError, Result,
};
use tokio::sync::RwLock;

use super::{
    StdioError,
    client::{Client, ClientHandle},
    launcher::Launcher,
};

#[derive(Serialize, Deserialize, Debug)]
pub struct StdioPluginConfig {
    pub command: String,
    pub args: Vec<String>,
    /// Environment variables to pass to the sub-process.
    #[serde(default, skip_serializing_if = "IndexMap::is_empty")]
    pub env: IndexMap<String, String>,
}

impl PluginConfig for StdioPluginConfig {
    type Error = StdioError;

    async fn create_plugin(
        self,
        working_directory: &std::path::Path,
        protocol_prefix: &str,
    ) -> error_stack::Result<Box<DynPlugin<'static>>, Self::Error> {
        // Look for the command in the system path and the working directory.
        let Self { command, args, env } = self;

        let launcher = Launcher::try_new(working_directory.to_owned(), command, args, env)?;

        Ok(DynPlugin::boxed(StdioPlugin::new(
            launcher,
            protocol_prefix.to_string(),
        )))
    }
}

pub struct StdioPlugin {
    state: RwLock<StdioPluginState>,
    protocol_prefix: String,
}

impl StdioPlugin {
    pub fn new(launcher: Launcher, protocol_prefix: String) -> Self {
        Self {
            state: RwLock::new(StdioPluginState::Uninitialized(launcher)),
            protocol_prefix,
        }
    }
}

enum StdioPluginState {
    Empty,
    Uninitialized(Launcher),
    Initialized(ClientHandle),
}

impl StdioPlugin {
    async fn client_handle(&self) -> Result<ClientHandle> {
        let guard = self.state.read().await;
        match &*guard {
            StdioPluginState::Initialized(handle) => Ok(handle.clone()),
            _ => Err(PluginError::Execution).attach_printable("stdio client not initialized"),
        }
    }

    async fn create_client(&self, context: Arc<dyn Context>) -> Result<ClientHandle> {
        let mut guard = self.state.write().await;
        match std::mem::replace(&mut *guard, StdioPluginState::Empty) {
            StdioPluginState::Uninitialized(launcher) => {
                let client = Client::try_new(launcher, context)
                    .await
                    .change_context(PluginError::Initializing)
                    .attach_printable("Unable to launch component server")?;
                *guard = StdioPluginState::Initialized(client.handle());
                Ok(client.handle())
            }
            _ => Err(PluginError::Initializing)
                .attach_printable_lazy(|| "Unexpected state".to_string()),
        }
    }
}

impl Plugin for StdioPlugin {
    async fn init(&self, context: &Arc<dyn Context>) -> Result<()> {
        let client = self.create_client(context.clone()).await?;

        client
            .request(&crate::schema::initialization::Request {
                runtime_protocol_version: 1,
                protocol_prefix: self.protocol_prefix.clone(),
            })
            .await
            .change_context(PluginError::Initializing)?;

        client
            .notify(&crate::schema::initialization::Complete {})
            .await
            .change_context(PluginError::Initializing)?;

        Ok(())
    }

    async fn list_components(&self) -> Result<Vec<Component>> {
        let client_handle = self.client_handle().await?;
        let response = client_handle
            .request(&crate::schema::list_components::Request {})
            .await
            .change_context(PluginError::ComponentInfo)?;

        Ok(response.components)
    }

    async fn component_info(&self, component: &Component) -> Result<ComponentInfo> {
        // TODO: Enrich this? Component not found, etc. based on the protocol error code?
        let client_handle = self.client_handle().await?;
        let response = client_handle
            .request(&crate::schema::component_info::Request {
                component: component.clone(),
            })
            .await
            .change_context(PluginError::ComponentInfo)?;

        Ok(response.info)
    }

    async fn execute(
        &self,
        component: &Component,
        context: ExecutionContext,
        input: ValueRef,
    ) -> Result<FlowResult> {
        let client_handle = self.client_handle().await?;
        let response = client_handle
            .request(&crate::schema::component_execute::Request {
                component: component.clone(),
                input,
                execution_id: context.execution_id(),
                step_id: context.step_id().unwrap_or("unknown").to_string(),
            })
            .await
            .change_context(PluginError::Execution)?;

        tracing::debug!("StdioPlugin: Received response: {:?}", response.output);

        // Check if the response contains a FlowResult by looking for the "outcome" field
        if let Some(outcome_obj) = response.output.as_object() {
            if let Some(outcome_value) = outcome_obj.get("outcome") {
                if let Some(outcome_str) = outcome_value.as_str() {
                    tracing::debug!("StdioPlugin: Found outcome field with value: {}", outcome_str);
                    // If the response has an "outcome" field, try to deserialize it as a FlowResult
                    match outcome_str {
                        "streaming" => {
                            // Try to deserialize as FlowResult::Streaming
                            tracing::debug!("StdioPlugin: Attempting to deserialize streaming result");
                            match serde_json::from_value::<stepflow_core::FlowResult>(response.output.as_ref().clone()) {
                                Ok(flow_result) => {
                                    tracing::info!("StdioPlugin: Successfully deserialized streaming result");
                                    return Ok(flow_result);
                                }
                                Err(e) => {
                                    tracing::warn!("Failed to deserialize streaming result: {}, attempting flexible construction", e);
                                    // Try to flexibly construct FlowResult::Streaming by adapting the structure
                                    if let Some(obj) = response.output.as_ref().as_object() {
                                        // Try to extract core streaming fields with flexible typing
                                        let stream_id = obj.get("stream_id")
                                            .and_then(|v| v.as_str())
                                            .unwrap_or("unknown");

                                        let chunk = obj.get("chunk")
                                            .and_then(|v| v.as_str())
                                            .unwrap_or("");

                                        let chunk_index = obj.get("chunk_index")
                                            .and_then(|v| v.as_u64().or_else(|| v.as_i64().map(|i| i as u64)))
                                            .unwrap_or(0) as usize;

                                        let is_final = obj.get("is_final")
                                            .and_then(|v| v.as_bool())
                                            .unwrap_or(false);

                                        // If we have a metadata field, use it; otherwise create one from remaining fields
                                        let metadata = if let Some(existing_metadata) = obj.get("metadata") {
                                            stepflow_core::workflow::ValueRef::new(existing_metadata.clone())
                                        } else {
                                            // Create metadata from all fields except the core streaming ones
                                            let mut metadata_obj = serde_json::Map::new();
                                            for (key, value) in obj.iter() {
                                                if !matches!(key.as_str(), "outcome" | "stream_id" | "chunk" | "chunk_index" | "is_final") {
                                                    metadata_obj.insert(key.clone(), value.clone());
                                                }
                                            }
                                            stepflow_core::workflow::ValueRef::new(serde_json::Value::Object(metadata_obj))
                                        };

                                        let flow_result = stepflow_core::FlowResult::Streaming {
                                            stream_id: stream_id.to_string(),
                                            metadata,
                                            chunk: chunk.to_string(),
                                            chunk_index,
                                            is_final,
                                        };

                                        tracing::info!("StdioPlugin: Successfully constructed streaming result flexibly");
                                        return Ok(flow_result);
                                    }
                                    tracing::warn!("Failed to flexibly construct streaming result, falling back to Success");
                                }
                            }
                        }
                        "success" | "failed" | "skipped" => {
                            // Try to deserialize as any FlowResult variant
                            match serde_json::from_value::<stepflow_core::FlowResult>(response.output.as_ref().clone()) {
                                Ok(flow_result) => {
                                    tracing::debug!("StdioPlugin: Successfully deserialized FlowResult: {:?}", flow_result);
                                    return Ok(flow_result);
                                }
                                Err(e) => {
                                    tracing::warn!("Failed to deserialize FlowResult: {}, falling back to Success", e);
                                }
                            }
                        }
                        _ => {
                            // Unknown outcome, treat as regular success
                            tracing::debug!("StdioPlugin: Unknown outcome '{}', treating as regular success", outcome_str);
                        }
                    }
                } else {
                    tracing::debug!("StdioPlugin: outcome field is not a string");
                }
            } else {
                tracing::debug!("StdioPlugin: No outcome field found in response");
            }
        } else {
            tracing::debug!("StdioPlugin: Response output is not an object");
        }

        // Default behavior: wrap in Success
        Ok(FlowResult::Success {
            result: response.output,
        })
    }
}
