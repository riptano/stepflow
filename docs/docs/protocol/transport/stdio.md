---
sidebar_position: 2
---

# STDIO Transport

The STDIO transport provides process-based communication between the Stepflow runtime and component servers using standard input and output streams. This transport is ideal for local component servers that can be executed as subprocesses.

## Overview

STDIO transport creates a subprocess for each component server and communicates via:
- **stdin**: Stepflow runtime sends JSON-RPC messages to the component server
- **stdout**: Component server sends JSON-RPC responses and bidirectional requests
- **stderr**: Component server sends logs and diagnostic information (not part of protocol)

:::warning
Relying on stderr for logging is discouraged, since it is not supported by other transports.
It may be useful in development or when accessing the component-server logs is easy, but in general a structrued logging mechanism (e.g., tracing) is preferred.
:::

## Process Lifecycle

The main benefit of the STDIO transport for local development is that the Stepflow runtime manages the component server processes.

### Process Creation
The Stepflow runtime spawns component server processes using the configured command and arguments:

See [Configuration Documentation](../../configuration/) for complete plugin configuration options.

### Process Management
- **Startup**: Runtime spawns process and waits for initialization
- **Health monitoring**: Runtime monitors process status and stdio streams
- **Graceful shutdown**: Runtime closes stdin and waits for process termination
- **Forced termination**: Runtime kills process after timeout if needed

### Process Environment
Process environment, working directory, and resource limits are configured via the runtime configuration. See [Configuration - STDIO Transport](../../configuration.md#stdio-transport) for details.