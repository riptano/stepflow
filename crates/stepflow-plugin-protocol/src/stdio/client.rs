use std::{ffi::OsString, path::PathBuf};

use error_stack::ResultExt as _;
use serde::de::DeserializeOwned;

use stepflow_protocol::{Method, Notification, Request};
use tokio::{
    sync::{mpsc, oneshot},
    task::JoinHandle,
};
use tracing::Instrument as _;
use uuid::Uuid;

use super::recv_message_loop::recv_message_loop;
use crate::incoming::OwnedIncoming;
use crate::stdio::{Result, StdioError};

/// Manages a client process spawned from a command.
///
/// Messages may be sent (as lines) to a channel that are sent via stdio.
pub struct Client {
    outgoing_tx: mpsc::Sender<String>,
    pending_tx: mpsc::Sender<(Uuid, oneshot::Sender<OwnedIncoming>)>,
    // TODO: Use the handle. We should actually check it for errors
    // before bubbling other (less meaningful) errors up.
    #[allow(dead_code)]
    loop_handle: JoinHandle<Result<()>>,
}

impl Client {
    pub fn builder(command: impl Into<PathBuf>) -> Builder {
        Builder::new(command)
    }

    pub async fn try_new(
        command: PathBuf,
        args: Vec<OsString>,
        working_directory: PathBuf,
    ) -> Result<Self> {
        error_stack::ensure!(command.is_file(), StdioError::InvalidCommand(command));

        let (outgoing_tx, outgoing_rx) = mpsc::channel(100);
        let (pending_tx, pending_rx) = mpsc::channel(100);

        let recv_span = tracing::info_span!("recv_message_loop", command = ?command, args = ?args);
        let loop_handle = tokio::spawn(
            recv_message_loop(command, args, working_directory, outgoing_rx, pending_rx)
                .instrument(recv_span),
        );

        Ok(Self {
            outgoing_tx,
            pending_tx,
            loop_handle,
        })
    }

    pub fn handle(&self) -> ClientHandle {
        ClientHandle {
            outgoing_tx: self.outgoing_tx.clone(),
            pending_tx: self.pending_tx.clone(),
        }
    }
}

pub struct ClientHandle {
    outgoing_tx: mpsc::Sender<String>,
    pending_tx: mpsc::Sender<(Uuid, oneshot::Sender<OwnedIncoming>)>,
}

impl ClientHandle {
    pub async fn request<I>(&self, params: &I) -> Result<I::Response>
    where
        I: Method + serde::Serialize + Send + Sync,
        I::Response: DeserializeOwned + Send + Sync + 'static,
    {
        let response = self.request_dyn(I::METHOD_NAME, params).await?;
        let raw_result = response.raw_value()?;
        let result: I::Response =
            serde_json::from_str(raw_result.get()).change_context(StdioError::InvalidResponse)?;
        Ok(result)
    }

    pub async fn notify<I>(&self, params: &I) -> Result<()>
    where
        I: Notification + serde::Serialize + Send + Sync,
    {
        self.send(&Request {
            jsonrpc: "2.0",
            id: None,
            method: I::NOTIFICATION_NAME,
            params,
        })
        .await?;

        Ok(())
    }

    async fn send(&self, msg: &(dyn erased_serde::Serialize + Send + Sync)) -> Result<()> {
        let msg = serde_json::to_string(&msg).change_context(StdioError::Send)?;

        self.outgoing_tx
            .send(msg)
            .await
            .change_context(StdioError::Send)?;

        Ok(())
    }

    async fn request_dyn(
        &self,
        method: &str,
        params: &(dyn erased_serde::Serialize + Send + Sync),
    ) -> Result<OwnedIncoming> {
        let id = Uuid::new_v4();
        let (response_tx, response_rx) = tokio::sync::oneshot::channel();

        self.pending_tx
            .send((id, response_tx))
            .await
            .map_err(|_| StdioError::Send)?;

        let request = Request {
            jsonrpc: "2.0",
            id: Some(id),
            method,
            params,
        };
        self.send(&request).await?;

        let response = response_rx.await.change_context(StdioError::Recv)?;
        debug_assert_eq!(response.id, Some(id));
        Ok(response)
    }
}

pub struct Builder {
    command: PathBuf,
    args: Vec<OsString>,
    working_directory: PathBuf,
}

impl Builder {
    pub(crate) fn new(command: impl Into<PathBuf>) -> Self {
        Self {
            command: command.into(),
            args: Vec::new(),
            working_directory: std::env::current_dir()
                .unwrap_or_else(|_| panic!("Failed to get current directory")),
        }
    }

    pub fn arg(mut self, arg: impl Into<OsString>) -> Self {
        self.args.push(arg.into());
        self
    }

    pub fn args<I: Into<OsString>>(mut self, args: impl IntoIterator<Item = I>) -> Self {
        for arg in args {
            self.args.push(arg.into());
        }
        self
    }

    pub fn working_directory(mut self, working_directory: impl Into<PathBuf>) -> Self {
        self.working_directory = working_directory.into();
        self
    }

    pub async fn build(self) -> Result<Client> {
        let client = Client::try_new(self.command, self.args, self.working_directory).await?;
        Ok(client)
    }
}
