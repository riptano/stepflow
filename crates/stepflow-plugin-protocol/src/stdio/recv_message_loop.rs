use std::{collections::HashMap, ffi::OsString, path::PathBuf, process::Stdio};

use error_stack::ResultExt as _;
use tokio::{
    io::{AsyncBufReadExt as _, AsyncWriteExt as _, BufReader},
    process::{Child, ChildStdin, ChildStdout},
    sync::{
        mpsc::{self, error::TryRecvError},
        oneshot,
    },
};
use tokio_stream::{StreamExt as _, wrappers::LinesStream};
use uuid::Uuid;

use crate::incoming::OwnedIncoming;
use crate::stdio::{Result, StdioError};

struct ReceiveMessageLoop {
    child: Child,
    to_child: ChildStdin,
    from_child: LinesStream<BufReader<ChildStdout>>,
    pending_requests: HashMap<Uuid, oneshot::Sender<OwnedIncoming>>,
}

impl ReceiveMessageLoop {
    fn try_new(command: PathBuf, args: Vec<OsString>, working_directory: PathBuf) -> Result<Self> {
        let child = tokio::process::Command::new(&command)
            .current_dir(&working_directory)
            .args(&args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .kill_on_drop(true)
            .spawn();

        let mut child = match child {
            Ok(child) => child,
            Err(e) => {
                tracing::error!(
                    "Failed to spawn child process '{} {args:?}': {e}",
                    command.display()
                );
                return Err(StdioError::Spawn.into());
            }
        };

        let to_child = child.stdin.take().expect("stdin requested");
        let from_child = child.stdout.take().expect("stdout requested");
        let from_child = LinesStream::new(BufReader::new(from_child).lines());

        Ok(Self {
            child,
            to_child,
            from_child,
            pending_requests: HashMap::new(),
        })
    }

    fn check_child_status(&mut self) -> Result<()> {
        if let Some(status) = self.child.try_wait().change_context(StdioError::Spawn)? {
            if !status.success() {
                tracing::error!("Child process exited with status {status}");
                return Err(StdioError::Spawn.into());
            }
        }
        Ok(())
    }

    async fn iteration(
        &mut self,
        outgoing_rx: &mut mpsc::Receiver<String>,
        pending_rx: &mut mpsc::Receiver<(Uuid, oneshot::Sender<OwnedIncoming>)>,
    ) -> Result<bool> {
        tokio::select! {
            child = self.child.wait() => {
                match child {
                    Ok(status) if status.success() => {
                        tracing::info!("Child process exited with status {status}");
                    }
                    Ok(status) => {
                        tracing::error!("Child process exited with status {status}");
                    }
                    Err(e) => {
                        tracing::error!("Child process exited with error: {e}");
                    }
                }
                Ok(false)
            }
            Some(outgoing) = outgoing_rx.recv() => {
                tracing::info!("Sending message to child: {outgoing}");
                self.to_child.write_all(outgoing.as_bytes()).await.change_context(StdioError::Send)?;
                self.to_child.write_all(b"\n").await.change_context(StdioError::Send)?;
                Ok(true)
            }
            Some(line) = self.from_child.next() => {
                let line = line.change_context(StdioError::Recv)?;
                tracing::info!("Received line from child: {line:?}");
                let msg = OwnedIncoming::try_new(line).change_context(StdioError::Recv)?;
                if let Some(id) = msg.id.as_ref() {
                    // This has an ID, so it's a method response.
                    if let Some(pending) = self.pending_requests.remove(id) {
                        // We've already seen the pending request, so send the result.
                        pending.send(msg).map_err(|_| StdioError::Send)?;
                        Ok(true)
                    } else {
                        // We haven't seen the pending request, so we'll receive from
                        // the pending_rx channel until we find it.
                        loop {
                            match pending_rx.try_recv() {
                                Ok((pending_id, pending_request)) => {
                                    if pending_id == *id {
                                        pending_request.send(msg).map_err(|_| StdioError::Send)?;
                                        return Ok(true);
                                    }
                                    self.pending_requests.insert(pending_id, pending_request);
                                }
                                Err(TryRecvError::Empty) => {
                                    // No more pending requests. This means the response we got
                                    // is unexpected. We'll log it and move on.
                                    tracing::warn!("Unexpected response: {msg:?}");
                                    return Ok(true);
                                }
                                Err(TryRecvError::Disconnected) => {
                                    // The pending_rx channel is closed, so we'll exit.
                                    tracing::warn!("Pending channel is closed.")
                                }
                            }
                        }
                    }
                } else {
                    // This is a notification.
                    todo!("handle notification {msg:?}");
                }
            }
            else => {
                tracing::info!("Exiting recv loop");
                Ok(false)
            }
        }
    }
}

pub async fn recv_message_loop(
    command: PathBuf,
    args: Vec<OsString>,
    working_directory: PathBuf,
    mut outgoing_rx: mpsc::Receiver<String>,
    mut pending_rx: mpsc::Receiver<(Uuid, oneshot::Sender<OwnedIncoming>)>,
) -> Result<()> {
    let mut recv_loop = ReceiveMessageLoop::try_new(command, args, working_directory)?;

    loop {
        match recv_loop.iteration(&mut outgoing_rx, &mut pending_rx).await {
            Ok(true) => {
                // Continue the loop.
            }
            Ok(false) => {
                // Exit the loop.
                break;
            }
            Err(mut e) => {
                tracing::info!("Error in recv loop: {e:?}. Checking child status.");
                if let Err(child_error) = recv_loop.check_child_status() {
                    e.extend_one(child_error);
                }
                return Err(StdioError::RecvLoop.into());
            }
        }
    }

    Ok(())
}
