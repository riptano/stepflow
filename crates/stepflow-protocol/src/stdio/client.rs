use std::sync::Arc;

use error_stack::ResultExt as _;
use serde::de::DeserializeOwned;
use stepflow_plugin::Context;

use crate::schema::{Method, Notification, RequestMessage};
use tokio::{
    sync::{mpsc, oneshot},
    task::JoinHandle,
};
use tracing::Instrument as _;
use uuid::Uuid;

use super::{launcher::Launcher, recv_message_loop::recv_message_loop};
use crate::incoming::OwnedIncoming;
use crate::stdio::{Result, StdioError};

/// Manages a client process spawned from a command.
///
/// Messages may be sent (as lines) to a channel that are sent via stdio.
pub struct Client {
    outgoing_tx: mpsc::Sender<String>,
    pending_tx: mpsc::Sender<(Uuid, oneshot::Sender<OwnedIncoming>)>,
    // TODO: Use the handle. We should actually check it for errors
    // before bubbling other (less meaningful) errors up.
    #[allow(dead_code)]
    loop_handle: JoinHandle<Result<()>>,
}

impl Client {
    pub async fn try_new(launcher: Launcher, context: Arc<dyn Context>) -> Result<Self> {
        let (outgoing_tx, outgoing_rx) = mpsc::channel(100);
        let (pending_tx, pending_rx) = mpsc::channel(100);

        let recv_span = tracing::info_span!("recv_message_loop", command = ?launcher.command, args = ?launcher.args);
        let loop_handle = tokio::spawn(
            recv_message_loop(
                launcher,
                outgoing_tx.clone(),
                outgoing_rx,
                pending_rx,
                context,
            )
            .instrument(recv_span),
        );

        Ok(Self {
            outgoing_tx,
            pending_tx,
            loop_handle,
        })
    }

    pub fn handle(&self) -> ClientHandle {
        ClientHandle {
            outgoing_tx: self.outgoing_tx.clone(),
            pending_tx: self.pending_tx.clone(),
        }
    }
}

#[derive(Clone)]
pub struct ClientHandle {
    outgoing_tx: mpsc::Sender<String>,
    pending_tx: mpsc::Sender<(Uuid, oneshot::Sender<OwnedIncoming>)>,
}

impl ClientHandle {
    pub async fn request<I>(&self, params: &I) -> Result<I::Response>
    where
        I: Method + serde::Serialize + Send + Sync,
        I::Response: DeserializeOwned + Send + Sync + 'static,
    {
        let response = self.request_dyn(I::METHOD_NAME, params).await?;
        let raw_result = response.raw_value()?;
        let result: I::Response =
            serde_json::from_str(raw_result.get()).change_context(StdioError::InvalidResponse)?;
        Ok(result)
    }

    pub async fn notify<I>(&self, params: &I) -> Result<()>
    where
        I: Notification + serde::Serialize + Send + Sync,
    {
        self.send(&RequestMessage {
            jsonrpc: "2.0",
            id: None,
            method: I::NOTIFICATION_NAME,
            params,
        })
        .await?;

        Ok(())
    }

    async fn send(&self, msg: &(dyn erased_serde::Serialize + Send + Sync)) -> Result<()> {
        let msg = serde_json::to_string(&msg).change_context(StdioError::Send)?;

        self.outgoing_tx
            .send(msg)
            .await
            .change_context(StdioError::Send)?;

        Ok(())
    }

    async fn request_dyn(
        &self,
        method: &str,
        params: &(dyn erased_serde::Serialize + Send + Sync),
    ) -> Result<OwnedIncoming> {
        let id = Uuid::new_v4();
        let (response_tx, response_rx) = tokio::sync::oneshot::channel();

        self.pending_tx
            .send((id, response_tx))
            .await
            .map_err(|_| StdioError::Send)?;

        let request = RequestMessage {
            jsonrpc: "2.0",
            id: Some(id),
            method,
            params,
        };
        self.send(&request).await?;

        let response = response_rx.await.change_context(StdioError::Recv)?;
        debug_assert_eq!(response.id, Some(id));
        Ok(response)
    }
}
