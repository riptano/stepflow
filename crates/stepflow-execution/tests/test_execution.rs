use std::fs::File;
use std::io::BufReader;

use error_stack::ResultExt as _;
use serde::Deserialize;
use stepflow_execution::execute;
use stepflow_plugin::{Plugin, Plugins};
use stepflow_plugin_protocol::stdio::{Client, StdioPlugin};
use stepflow_plugin_testing::{MockComponentBehavior, MockPlugin};
use stepflow_workflow::Value;
use tracing_subscriber::{EnvFilter, layer::SubscriberExt as _, util::SubscriberInitExt as _};

static INIT_TEST_LOGGING: std::sync::Once = std::sync::Once::new();

/// Makes sure logging is initialized for test.
///
/// This needs to be called on each test.
pub fn init_test_logging() {
    INIT_TEST_LOGGING.call_once(|| {
        // We don't use a test writer for end to end testts.
        let fmt_layer = tracing_subscriber::fmt::layer();

        tracing_subscriber::registry()
            .with(EnvFilter::new("stepflow_=trace,info"))
            .with(fmt_layer)
            .with(tracing_error::ErrorLayer::default())
            .try_init()
            .unwrap();
    });
}

#[derive(Deserialize)]
struct TestFlow {
    #[serde(flatten)]
    flow: stepflow_workflow::Flow,
    test_cases: Vec<TestCase>,
}

#[derive(Deserialize)]
struct TestCase {
    input: serde_json::Value,
    #[serde(default)]
    expect_failure: bool,
}

fn create_mock_plugin() -> MockPlugin {
    let mut mock_plugin = MockPlugin::new("mock");
    mock_plugin
        .mock_component("mock://one_output")
        .behavior(
            Value::new(serde_json::json!({ "input": "a" })),
            MockComponentBehavior::Valid {
                output: Value::new(serde_json::json!({ "output": "b" })),
            },
        )
        .behavior(
            Value::new(serde_json::json!({ "input": "hello" })),
            MockComponentBehavior::Valid {
                output: Value::new(serde_json::json!({ "output": "world" })),
            },
        );
    mock_plugin
        .mock_component("mock://two_outputs")
        .behavior(
            Value::new(serde_json::json!({ "input": "b" })),
            MockComponentBehavior::Valid {
                output: Value::new(serde_json::json!({ "x": 1, "y": 2 })),
            },
        )
        .behavior(
            Value::new(serde_json::json!({ "input": "world" })),
            MockComponentBehavior::Valid {
                output: Value::new(serde_json::json!({ "x": 2, "y": 8 })),
            },
        );
    mock_plugin
}

/// Simple error we can wrap any other errors in.
#[derive(Debug, thiserror::Error)]
#[error("test error")]
struct TestError;

async fn create_python_plugin() -> error_stack::Result<StdioPlugin, TestError> {
    let uv = which::which("uv").change_context(TestError)?;
    tracing::info!("Found uv at {uv:?}");

    // Determine the path to the `sdks/python` directory from the CARGO_MANIFEST_DIR.
    let manifest_dir = std::env::var("CARGO_MANIFEST_DIR").change_context(TestError)?;
    let python_dir = std::path::Path::new(&manifest_dir)
        .parent()
        .ok_or(TestError)?
        .parent()
        .ok_or(TestError)?
        .join("sdks")
        .join("python");
    let python_dir = python_dir.to_str().ok_or(TestError)?;

    let client = Client::builder(uv)
        .args(["--project", python_dir, "run", "stepflow_sdk"])
        .build()
        .await
        .change_context(TestError)?;

    let plugin = StdioPlugin::new(client.handle());
    plugin.init().await.change_context(TestError)?;
    Ok(plugin)
}

fn run_tests(plugins: Plugins, rt: tokio::runtime::Handle) {
    insta::glob!("flows/*.yaml", |path| {
        rt.block_on(async {
            let file = File::open(path).unwrap();
            let reader = BufReader::new(file);
            let TestFlow { flow, test_cases } =
                serde_yml::from_reader::<_, TestFlow>(reader).unwrap();

            let mut settings = insta::Settings::clone_current();
            settings.set_input_file(path);

            for (index, test_case) in test_cases.into_iter().enumerate() {
                settings.set_description(format!("case {index}"));
                settings.set_info(&test_case.input);

                let result = execute(&plugins, &flow, Value::new(test_case.input)).await;
                if test_case.expect_failure {
                    let result = result.unwrap_err();
                    settings.bind(|| {
                        insta::assert_yaml_snapshot!(result);
                    });
                } else {
                    let result = result.unwrap();
                    settings.bind(|| {
                        insta::assert_yaml_snapshot!(result);
                    });
                }
            }
        })
    });
}

#[tokio::test(flavor = "multi_thread")]
async fn execute_flows() {
    init_test_logging();

    let mut plugins = Plugins::new();

    plugins.register("mock".to_owned(), create_mock_plugin());
    plugins.register("python".to_owned(), create_python_plugin().await.unwrap());

    let rt = tokio::runtime::Handle::current();

    let handle = rt.clone().spawn_blocking(|| run_tests(plugins, rt));
    handle.await.unwrap();
}
