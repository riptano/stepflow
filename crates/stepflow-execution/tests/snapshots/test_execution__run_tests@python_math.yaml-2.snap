---
source: crates/stepflow-execution/tests/test_execution.rs
description: case 1
expression: result
info:
  m: 8
  n: 10
input_file: crates/stepflow-execution/tests/flows/python_math.yaml
---
m_plus_n: 18
m_plus_n_times_n: 180
m_times_n: 80
